fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanCharge.java

```diff
@@ -494,17 +494,29 @@ private void updateInstallmentCharges() {
             this.loanInstallmentCharge.addAll(chargePerInstallments);
         } else {
             int index = 0;
-            final LoanInstallmentCharge[] loanChargePerInstallments = new LoanInstallmentCharge[chargePerInstallments.size()];
-            final LoanInstallmentCharge[] loanChargePerInstallmentArray = chargePerInstallments.toArray(loanChargePerInstallments);
-            for (final LoanInstallmentCharge chargePerInstallment : this.loanInstallmentCharge) {
+            //FINERACT-7: NB: this is really overkill and slow, without changing the Sets and or Entities this is the most
+            //locally safe way to fix this issue
+            ArrayList<LoanInstallmentCharge> oldChargeInstallments = new ArrayList<LoanInstallmentCharge>();
+            ArrayList<LoanInstallmentCharge> newChargeInstallments = new ArrayList<LoanInstallmentCharge>();
+
+            oldChargeInstallments.addAll(this.loanInstallmentCharge);
+            newChargeInstallments.addAll(chargePerInstallments);
+            Collections.sort(oldChargeInstallments);
+            Collections.sort(newChargeInstallments);
+            final LoanInstallmentCharge[] loanChargePerInstallmentArray = chargePerInstallments.toArray(new LoanInstallmentCharge[chargePerInstallments.size()]);
```

Comment:
I believe this array object is unnecessary. You can modify Loan.generateInstallmentLoanCharges(final LoanCharge loanCharge) method to return List object. Please use this list object